### PR TITLE
shellhub-agent: Add service and package expressions

### PR DESCRIPTION
**Pull request by otavio**
###### Motivation for this change

We aim to ease the use of ShellHub and NixOS is one of the distros we'd like to support. It is working very well on our machines.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/shellhub-agent/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/shellhub-agent/default.nix:11]:
```suggestion
    owner = "shellhub-io";
    repo = "shellhub";
    rev = "v${version}";
    sha256 = "0cd41ing1pcf1bdaaq00w5h7lih5j2kcaa0m41g3ikm3vd1w5qna";
```

File: pkgs/applications/networking/shellhub-agent/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/shellhub-agent/default.nix:23]: The text is pretty long and fits better into longDescription. Can you add a shorter one which does not exceed about 80 characters?

File: pkgs/applications/networking/shellhub-agent/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/shellhub-agent/default.nix:27]:
```suggestion
    maintainers = with maintainers; [ otavio ];
```

File: pkgs/applications/networking/shellhub-agent/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/shellhub-agent/default.nix:14]: It would be great if you could upstream those changes and then reference the patch from the commit hash in the PR. Like https://github.com/NixOS/nixpkgs/commit/51b455552402934e7a3abd49e843e46774078664.patch.

File: nixos/modules/services/networking/shellhub-agent.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/shellhub-agent.nix:54]:
```suggestion
      };
    };
  };
```

File: nixos/modules/services/networking/shellhub-agent.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/shellhub-agent.nix:82]:
```suggestion
      };
    };
```

File: nixos/modules/services/networking/shellhub-agent.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/shellhub-agent.nix:88]:
```suggestion
    environment.systemPackages = [ pkgs.shellhub-agent ];
  };
}
```

File: nixos/modules/services/networking/shellhub-agent.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/shellhub-agent.nix:9]:
```suggestion
with lib;
let
  cfg = config.services.shellhub-agent;
in {
```

File: nixos/modules/services/networking/shellhub-agent.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/shellhub-agent.nix:28]: What happens if someone uses this id? Would it just not work or do harm?

File: nixos/modules/services/networking/shellhub-agent.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/shellhub-agent.nix:45]: I am not sure if this path works well with NixOS. It would be great if someone else could verify if this is okay.

File: pkgs/applications/networking/shellhub-agent/default.nix
Comment (otavio) [reply pkgs/applications/networking/shellhub-agent/default.nix:23]: Sure; I'll rework it.

File: pkgs/applications/networking/shellhub-agent/default.nix
Comment (otavio) [reply pkgs/applications/networking/shellhub-agent/default.nix:14]: We are working on the 0.5.0 release and I'll update it once it is released, dropping this patch.

File: nixos/modules/services/networking/shellhub-agent.nix
Comment (otavio) [reply nixos/modules/services/networking/shellhub-agent.nix:25]: The server provides this, so if you use an invalid tenant Id, it won't connect on the server-side.

File: nixos/modules/services/networking/shellhub-agent.nix
Comment (otavio) [reply nixos/modules/services/networking/shellhub-agent.nix:42]: If there is a better way to do that, I am keen to learn and adapt.

**Comment by otavio** [comment]
@SuperSandro2000 I succeed in clean it. Please do another review on this when you have time. Thanks for all the comments.

File: pkgs/applications/networking/shellhub-agent/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/shellhub-agent/default.nix:32]: This is automatically set to ``linux ++ darwin`` in https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/go-modules/generic/default.nix#L239. This should be removed here and only be set if only linux or darwin is supported.

**Review by SuperSandro2000 — COMMENTED** [review]
I am not that familiar with modules so someone else should take a look at it.

File: pkgs/applications/networking/shellhub-agent/default.nix
Comment (otavio) [reply pkgs/applications/networking/shellhub-agent/default.nix:32]: Oh, good point. I fixed it and pushed it again.

**Comment by otavio** [comment]
Is there anything else pending for this PR?

File: maintainers/maintainer-list.nix
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:6566]: Please put this into a separate commit with the message ``maintainers: add otavio``.

File: pkgs/applications/networking/shellhub-agent/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/shellhub-agent/default.nix:22]:
```suggestion
      "Enables easy access any Linux device behind firewall and NAT";
```

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 104419` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 104419` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>shellhub-agent</li>
  </ul>
</details>

File: maintainers/maintainer-list.nix
Comment (otavio) [reply maintainers/maintainer-list.nix:6566]: This is already in a specific commit; I reworked the commit log as you suggested.

File: pkgs/applications/networking/shellhub-agent/default.nix
Comment (otavio) [reply pkgs/applications/networking/shellhub-agent/default.nix:22]: Applied.

**Comment by otavio** [comment]
Hello all, I rebased this on top of master to resolve the conflict. Could someone take a look at this?

File: pkgs/applications/networking/shellhub-agent/default.nix
Comment (jonringer) [inline pkgs/applications/networking/shellhub-agent/default.nix:6]: This should probably be a different commit:
```
shellhub-agent: init at 0.4.2
```

**Review by jonringer — COMMENTED** [review]
I would break out the package and service into 2 different ocmmits:
```
shellhub-agent: init at 0.4.2
nixos/shellhub-agent: init
```

File: nixos/modules/services/networking/shellhub-agent.nix
Comment (otavio) [reply nixos/modules/services/networking/shellhub-agent.nix]: Done.

File: pkgs/applications/networking/shellhub-agent/default.nix
Comment (otavio) [reply pkgs/applications/networking/shellhub-agent/default.nix:6]: Done.

File: nixos/modules/services/networking/shellhub-agent.nix
Comment (aanderse) [reply nixos/modules/services/networking/shellhub-agent.nix]: Or overlays. Don't forget those!

Comment (otavio) [reply nixos/modules/services/networking/shellhub-agent.nix]: Absolutely. I reworked the commits as suggested by @jonringer, so I think it is fine now. @aanderse, please take a look as well.

**Review by jonringer — APPROVED** [review]
LGTM

cc @Mic92  or @Infinisil for final look-over of module

File: nixos/modules/services/networking/shellhub-agent.nix
Comment (Infinisil) [inline nixos/modules/services/networking/shellhub-agent.nix:83]: This runs the service as root. Is this required? If it is, then I suggest explicitly setting `User = "root"` and adding a comment why it's necessary. It may also be possible to not require root, but just some special capabilities, in which case you could try `DynamicUser = true` and `AmbientCapabalities`/`CapabilityBoundingSet` (see other modules for examples).

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/shellhub-agent.nix
Comment (otavio) [reply nixos/modules/services/networking/shellhub-agent.nix:85]: Thanks @Infinisil for the review. For now, we need to use root but I'll investigate if in the future we can drop the capabilities.

Comment (Infinisil) [reply nixos/modules/services/networking/shellhub-agent.nix:85]: You added a comment mentioning that it requires root, but it doesn't explain _why_ that's the case, which is what I'm looking for (probably just because it needs to log into arbitrary users?)

Comment (otavio) [reply nixos/modules/services/networking/shellhub-agent.nix:85]: It is an SSH server so yes, it needs to initialize sessions with different users. I assumed this was implicit by it being an SSH server. Should I make this explicit there?

Comment (otavio) [reply nixos/modules/services/networking/shellhub-agent.nix:85]: I added this explicitly there.

**Review by infinisil — APPROVED** [review]
That's great, thanks :)

**Pull request merged**
